### PR TITLE
Feature inventory

### LINKED ISSUE #145: Entity Interaction
As the VR player, I want to be able to interact with entities, so that I can pickup bombs and keys or open doors.

This task has subtasks:
- #146 Entity Interaction - Inventory
- #147 Entity Interaction - Bombs
- #148 Entity Interaction - Keys
- #149 Entity Interaction - Doors

### LINKED ISSUE #146: Entity Interaction - Inventory
Add an inventory system to the player.

This task is a subtask of #145 Entity Interaction.

### LINKED ISSUE #147: Entity Interaction - Bombs
Implement the ability to pick up bombs and place bombs.

This is a subtask of #145 Entity Interaction.

### LINKED ISSUE #148: Entity Interaction - Keys
Implement the ability to pick up keys, which adds them to the inventory implemented in task #146.

This is a subtask of #145 Entity Interaction.

### LINKED ISSUE #149: Entity Interaction - Doors
Implement the ability to open doors using the correct key from the inventory (#146).

This is a subtask of #145 Entity Interaction.

### PR DESCRIPTION
**Pull request by WytzeElhorst**
Adds inventory class with inventory test with 97% line coverage as well as entity interaction between keys and door and the ability to pick up key and bombs and drop the latter.

this closes #145  closes #146  closes #147  closes #148 and  closes  #149

### CODE REVIEW

File: src/main/java/nl/tudelft/contextproject/model/VRPlayer.java [hunk 1]
```diff
@@ -169,15 +172,93 @@ public void onAction(String name, boolean isPressed, float tpf) {
...
+	}
 	
+	/**
+	 * Player picks up a nearby item.
+	 * Also opens nearby doors if possible
+	 */
+	public void pickUp() {
+		Vector3f vec = this.getSpatial().getLocalTranslation();
+		List<Entity> list = new ArrayList<Entity>();
+		list = Main.getInstance().getCurrentGame().getEntities();
```
Comment (Taeir) [inline src/main/java/nl/tudelft/contextproject/model/VRPlayer.java]: This is going to give trouble with the new changes made in #191.

File: src/test/java/nl/tudelft/contextproject/model/VRPlayerTest.java [hunk 2]
```diff
@@ -126,4 +127,26 @@ public void testGetSpatialCheckGravity() {
...
+	/**
+	 * Tests that the dropbomb method removes a bomb from your inventory.
+	 */
+	@Test 
+	public void testDropBomb() {
+		setupGeometryMock();
+		player.getInventory().add(new Bomb());
+		player.dropBomb();
+		assertTrue(player.getInventory().size() == 0);
+		}
```
Comment (Taeir) [inline src/test/java/nl/tudelft/contextproject/model/VRPlayerTest.java]: Indentation

File: src/test/java/nl/tudelft/contextproject/model/VRPlayerTest.java [hunk 3]
```diff
@@ -126,4 +127,26 @@ public void testGetSpatialCheckGravity() {
...
+	/**
+	 * tests that the dropbomb method doesn't remove a bomb when there is none.
+	 */
+	@Test
+	public void testDropnoBomb() {
+		setupGeometryMock();
+		player.getInventory().add(new Key(ColorRGBA.Yellow));
+		player.dropBomb();
+		assertTrue(player.getInventory().size() == 1);
+		}
```
Comment (Taeir) [inline src/test/java/nl/tudelft/contextproject/model/VRPlayerTest.java]: Indentation.

**Comment by Taeir** [comment]
Code looks good. There are a few minor things:
1) You use a lot of normal for loops over lists, it is probably better and less code to use for each loops (`for (Entity e : entities)`).
2) Indentation is wrong here and there.

File: src/main/java/nl/tudelft/contextproject/model/Bomb.java [hunk 4]
```diff
@@ -83,4 +84,17 @@ public void move(float x, float y, float z) {
...
+	 * Method to move the bomb.
+	 * @param x
+	 * 		The x coordinate of the bomb
+	 * @param y
+	 * 		The y coordinate of the bomb
+	 * @param z
+	 * 		The z coordinate of the bomb
+	 */
+	public void move(int x, int y, int z) {
+		sp.move(x, y, z);
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/Bomb.java]: This does not move the collision object. Also @override should be added.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	private int keys;
+	private int bombs;
+
+	/**
+	 * Constructor for the inventory, starts empty with 0 keys and doors.
+	 */
+	public Inventory() {
+		array = new ArrayList<Entity>();
+		keys = 0;
+		bombs = 0;
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java]: (nitpicking) this.array etc.

File: src/main/java/nl/tudelft/contextproject/model/Key.java [hunk 6]
```diff
@@ -12,28 +11,33 @@
...
 import com.jme3.scene.Spatial;
 import com.jme3.scene.shape.Box;
+
 import nl.tudelft.contextproject.Main;
 /**
  * Class representing a key.
  */
 public class Key extends Entity implements PhysicsObject {
 	private Geometry geometry;
 	private Spatial sp;
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/Key.java:20]: You might want to remove geometry

File: src/main/java/nl/tudelft/contextproject/model/VRPlayer.java [hunk 7]
```diff
@@ -169,15 +174,92 @@ public void onAction(String name, boolean isPressed, float tpf) {
...
+	/**
+	 * Player picks up a nearby item.
+	 * Also opens nearby doors if possible
+	 */
+	public void pickUp() {
+		Vector3f vec = this.getSpatial().getLocalTranslation();
+		Set<Entity> set = Main.getInstance().getCurrentGame().getEntities();
+		for (Entity ent : set) {
+			Vector3f vec2 = ent.getSpatial().getLocalTranslation();
+			if (Math.abs((int) vec.x - vec2.x) <= 2 && Math.abs((int) vec.y - vec2.y) <= 2 && Math.abs((int) vec.z - vec2.z) <= 2) {
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/VRPlayer.java]: replace with if(ent.collidesWithPlayer(2f))

**Comment by ChielBruin** [comment]
All looks good! I made some comments, so please address these or elaborate why it is better this way before merging

**Comment by ChielBruin** [comment]
Can be merged now in my opinion as the suggested changes are made.

**Comment by Geweldig** [comment]
Can be merged.

**Pull request merged**